### DISCUSSION TITLE
exporting all fig-alt together with figure numbers

### DISCUSSION THREAD

**Discussion by baptiste** [quarto-dev/quarto-cli#7526] [Q&A]
### Description

I'm writing a book with CRC using Quarto + knitr + Latex, and the publishers are asking that the ALT text for each figure be copied into a single Word document. Has anyone written a Lua filter doing something like this? 

- grab the `fig-alt` field of each figure chunk
- dump the text into a new Word document, together with the figure number

Many thanks

**Comment by mcanouil** [comment]
That should not be too hard.
Do you have a small reproducible example document to use?
(Note that likely to have a post-render script for Word document)

**Comment by cscheid** [comment]
This is "easy" in 1.4. You'll need:

1. A filter that targets FloatRefTarget nodes
2. ... and that runs _after_ crossrefs are handled. In 1.4 you have better control over when filters run. That's not currently documented (🤦) but you can see an example [here](https://github.com/quarto-dev/quarto-cli/tree/main/tests/docs/smoke-all/2023/09/28/7066.qmd). The currently available points for injecting a filter are: `pre-ast`, `post-ast` (or its synonym `pre-quarto`), `post-quarto` (or its synonym `pre-render`), and `post-render`.

`pre-quarto` and `post-render` are the old injection points of "pre" filters and "post" filters in Quarto 1.3 and earlier.

This separation clarifies the three major phases in our Lua filters: `ast` processing handles normalization of the input markdown into a semantic AST; `quarto` handles what quarto does: resolving citations, references, layouts, subfloats, etc, but does _not_ emit format-specific output. `render` takes the processed AST and produces a format-specific document.

(This separation is at least somewhat aspirational, and there's still a bit of format-specific output in the `quarto` phase, specifically in the resolution of references. We'll be cleaning this up progressively in future versions of quarto.)

## Specifics

Your filter will look more or less like this:

```lua
local captures = pandoc.Blocks({})
return {
  FloatRefTarget = function(float)
    -- float.order is a data structure that stores where the float is found
    -- use quarto.log.output() to inspect its contents, or follow the code in `refNumberOption`
    -- from src/resources/filters/crossref/format.lua
    -- then, use that to create a pandoc.Para
    captures:insert(pandoc.Para(...))
  end,
  Pandoc = function(doc)
    return pandoc.Pandoc(captures) -- replaces the result with a completely new document
  end,
}
```

Then, in your .qmd you'll use:

````
---
filters:
  - at: post-quarto
    path: capture.lua
---
````

Of course, you probably want to consider putting this metadata in a separate .yml file that you conditionally invoke, so that you get your regular content unless you specifically enable the filter.

This is the kind of thing we would love to support better in the future, and why we did the crossref overhaul. One of the things that's pretty janky right now is that we don't expose refNumberOption, etc, or document the `order` field at all. We'll be improving on this as time goes by, and we could eventually change those fields, but I don't expect we will anytime soon.

## EDIT: "Oh, a _book_"

This is significantly more complicated :( The fundamental problem is that book rendering is multi-file, and so the crossref information is computed differently. For books, we have an additional Pandoc pass that runs in formats that produce a single file (.epub, .pdf, .docx, etc) We'll still need to run the filter above, but only in the final pass that produces the single output.

If this weren't a user filter, you'd use `param("single-file-book", false)` to check in which pass the filter is getting called; but you don't have access to that :(

Let me talk to the other quarto folks and get back to you.

**Reply by baptiste** [reply]
Thanks a lot! Yeah, I figured the book aspect might complicate things... Looking forward to hearing more. 
~~I wonder if a custom pandoc Writer might be a way to go? Presumably at that stage quarto is done with its numbering etc., and is letting pandoc do its thing with the output format. The Writer could presumably just call the standard Writer but also output the fig-alt content into a new document in parallel?~~ (_Edit:_ nevermind, I guess that's similar to post-quarto or post-render, but still called in the multiple passes for split documents)

**Reply by baptiste** [reply]
By the way quarto v1.4.489 seems to not show `fig-alt` in html in [this example](https://mine-cetinkaya-rundel.github.io/quarto-tip-a-day/posts/28-fig-alt/). I thought it might be easiest for now to post-process a concrete output document (e.g. HTML, or whatever else keeps the ALT info in the output), grep the figure number and the ALT content, and dump those into a Word document converted via pandoc.
